src/coupled_driver.cpp

```diff
@@ -191,132 +226,213 @@ void CoupledDriver::compute_temperature_norm(const CoupledDriver::Norm& n,
+
+  auto& neutronics = this->get_neutronics_driver();
+  auto& heat = this->get_heat_driver();
+
+  // *********************************************************************
+  // Send heat driver's density to coupling root and apply underrelaxation
+  // *********************************************************************
+
+  if (relax && comm_.rank == coupling_root_) {
     std::copy(densities_.begin(), densities_.end(), densities_prev_.begin());
   }
 
-  auto& heat = get_heat_driver();
-  if (heat.active()) {
-    auto d = heat.density();
+  densities_ = heat.density();
+  comm_.send_and_recv(densities_, coupling_root_, heat_root_);
 
-    if (heat.has_coupling_data())
-      densities_ = alpha_rho_ * d + (1.0 - alpha_rho_) * densities_prev_;
+  if (relax && comm_.rank == coupling_root_) {
+    densities_ = alpha_rho_ * densities_ + (1.0 - alpha_rho_) * densities_prev_;
   }
 
-  // Set density in the neutronics solver
-  set_density();
-}
+  // ************************************************************************
+  // Send underrelaxed density to all neutron ranks and set cell temperatures
+  // ************************************************************************
 
-bool CoupledDriver::has_global_coupling_data() const
-{
-  return this->get_neutronics_driver().active() && this->get_heat_driver().active();
+  comm_.send_and_recv(densities_, neutronics_root_, coupling_root_);
+  neutronics.comm_.broadcast(densities_);
+
+  if (neutronics.active()) {
+    // For each neutronics cell in a fluid cell, volume average the densities
+    // and set in driver
+    for (CellHandle cell = 0; cell < cell_to_elems_.size(); ++cell) {
+      if (cell_fluid_mask_[cell] == 1) {
+        // Get volume-averaged density for this cell instance
+        double average_density = 0.0;
+        double total_vol = 0.0;
+        for (int e : cell_to_elems_.at(cell)) {
+          average_density += densities_[e] * elem_volumes_[e];
+          total_vol += elem_volumes_[e];
+        }
+        // Set density for cell instance
+        average_density /= total_vol;
+        Ensures(average_density > 0.0);
+        neutronics.set_density(cell, average_density);
+      }
+    }
+  }
 }
 
 void CoupledDriver::init_mappings()
 {
   comm_.message("Initializing mappings");
 
   const auto& heat = this->get_heat_driver();
-  if (heat.active()) {
-    // Get centroids from heat driver
-    auto elem_centroids = heat.centroids();
-
-    // Broadcast centroids onto all the neutronics procs
-    this->get_neutronics_driver().comm_.broadcast(elem_centroids);
+  auto& neutronics = this->get_neutronics_driver();
 
-    // Set element->cell and cell->element mappings. Create buffer to store cell
-    // handles corresponding to each heat-fluids global element.
-    elem_to_cell_.resize(heat.n_global_elem());
+  std::vector<Position> elem_centroids;
+  if (comm_.rank == heat_root_ || neutronics.active()) {
+    elem_centroids.resize(n_global_elem_);
+  }
```

Comment:
Doesn't seem like we need to resize this guy up. `elem_centroids = heat.centroids()` will give the proper size on `heat_root_`, and the broadcast on the neutronics comm will take care of the size for all neutronics ranks.

Reply:
Fixed in https://github.com/enrico-dev/enrico/pull/93/commits/324c4a531606441db6394cf3998c9d805c9d5c3a